src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestUpdate.java

```diff
@@ -0,0 +1,2161 @@
+        }
+
+        try {
+            w.igit().fixSubmoduleUrls("origin", listener);
+            assertTrue("not CliGit", w.igit() instanceof CliGitAPIImpl);
+        } catch (UnsupportedOperationException uoe) {
+            assertTrue("Unsupported operation not on JGit", w.igit() instanceof JGitAPIImpl);
+        }
+    }
+
+
+    /* Shows the JGit submodule update is broken now that tests/getSubmodule includes a renamed submodule */
+    @Test
+    public void testGetSubmodules() throws Exception {
+        w.init();
+        w.git.clone_().url(localMirror()).repositoryName("sub_origin").execute();
+        w.git.checkout().ref("sub_origin/tests/getSubmodules").branch("tests/getSubmodules").execute();
+        List<IndexEntry> r = w.git.getSubmodules("HEAD");
+        assertEquals(
+                "[IndexEntry[mode=160000,type=commit,file=modules/firewall,object=978c8b223b33e203a5c766ecf79704a5ea9b35c8], " +
+                        "IndexEntry[mode=160000,type=commit,file=modules/ntp,object=b62fabbc2bb37908c44ded233e0f4bf479e45609], " +
+                        "IndexEntry[mode=160000,type=commit,file=modules/sshkeys,object=689c45ed57f0829735f9a2b16760c14236fe21d9]]",
+                r.toString()
+        );
+        w.git.submoduleInit();
+        w.git.submoduleUpdate().execute();
+
+        assertTrue("modules/firewall does not exist", w.exists("modules/firewall"));
+        assertTrue("modules/ntp does not exist", w.exists("modules/ntp"));
+        // JGit submodule implementation doesn't handle renamed submodules
+        if (w.igit() instanceof CliGitAPIImpl) {
+            assertTrue("modules/sshkeys does not exist", w.exists("modules/sshkeys"));
+        }
+        assertFixSubmoduleUrlsThrows();
+    }
+
+
+    /* Shows the submodule update is broken now that tests/getSubmodule includes a renamed submodule */
+    @NotImplementedInJGit
+    @Test
+    public void testSubmoduleUpdate() throws Exception {
+        w.init();
+        w.git.clone_().url(localMirror()).repositoryName("sub2_origin").execute();
+        w.git.checkout().branch("tests/getSubmodules").ref("sub2_origin/tests/getSubmodules").deleteBranchIfExist(true).execute();
+        w.git.submoduleInit();
+        w.git.submoduleUpdate().execute();
+
+        assertTrue("modules/firewall does not exist", w.exists("modules/firewall"));
+        assertTrue("modules/ntp does not exist", w.exists("modules/ntp"));
+        // JGit submodule implementation doesn't handle renamed submodules
+        if (w.igit() instanceof CliGitAPIImpl) {
+            assertTrue("modules/sshkeys does not exist", w.exists("modules/sshkeys"));
+        }
+        assertFixSubmoduleUrlsThrows();
+
+        String shallow = Paths.get(".git", "modules", "module", "1", "shallow").toString();
+        assertFalse("shallow file existence: " + shallow, w.exists(shallow));
+    }
+
+    @NotImplementedInJGit
```

Comment:
Annotation is not honored by JUnit 4 implementation.  Remove and either move the test to a CliGit specific file or guard the test with a conditional
```suggestion
```